File: models/heartbeat.py
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import DateTime, func
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base


class Heartbeat(Base):
    __tablename__ = "heartbeats"
    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    # user_id: Mapped[str] = mapped_column(ForeignKey(User.id), unique=True)
    user_id: Mapped[str] = mapped_column(index=True)
    entity: Mapped[str] = mapped_column(index=True, nullable=True)
    type: Mapped[str] = mapped_column(index=True, nullable=True)
    category: Mapped[str] = mapped_column(index=True, nullable=True)
    browser: Mapped[str] = mapped_column(index=True)

    domain: Mapped[str] = mapped_column(index=True)  # page domain
    path: Mapped[str] = mapped_column(index=True)  # page path
    user_agent: Mapped[str] = mapped_column()  # user agent

    time: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), index=True, nullable=True
    )
    hash: Mapped[str] = mapped_column(index=True, unique=True, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
    )

File: etl/src/process/entity/heartbeat.rs
use sea_orm::FromQueryResult;

use super::*;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "heartbeat")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: Uuid,
    pub user_id: Uuid,
    pub entity: String,
    #[sea_orm(column_name = "type")]
    pub kind: String,
    pub category: String,
    pub browser: String,
    pub domain: String,
    pub path: String,
    pub user_agent: String,
    pub time: Time,
    pub hash: String,
    pub created_at: Time,
}

#[derive(DerivePartialModel, FromQueryResult)]
#[sea_orm(entity = "Entity")]
pub struct PartialModel {
    #[sea_orm(primary_key)]
    pub id: Uuid,
    pub user_id: Uuid,
    pub domain: String,
    pub path: String,
    pub time: chrono::NaiveDateTime,
    pub hash: String,
    pub created_at: chrono::NaiveDateTime,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}
